File: 35.py

def is_prime_basic(x):
    if x <= 1:
        return False
    elif x <= 3:
        return True
    elif x % 2 == 0 or x % 3 == 0:
        return False
    i = 5
    while i ** 2 <= x:
        if x % i == 0 or x % (i+2) == 0:
            return False
        i += 6
    return True


def rotations(digits):
    l = []
    for i in range(len(digits) - 1):
        digits = [digits[-1]] + digits[:-1]
        l.append(int("".join(map(str, digits))))
    return l


def is_circular_prime(p):
    digits = list(map(int, str(p)))
    if sum([1 if x % 2 == 0 else 0 for x in digits]) != 0 and p != 2:
        return False
    combinations = rotations(digits)
    for x in combinations:
        if not is_prime_basic(x):
            return False
    return True


c_p = 0
for i in range(1000000):
    print(i, "\r", end="")
    if is_prime_basic(i) and is_circular_prime(i):
            c_p += 1

print()
print(c_p)

File: cpp/35.cpp
#include <iostream>
#include <vector>

using namespace std;

bool is_prime(int n){
    if (n <= 1)
    {
        return false;
    } else if (n <= 3)
    {
        return true;
    } else if (n % 2 == 0 || n % 3 == 0)
    {
        return false;
    }

    int i =5;

    while (i * i <= n)
    {
        if (n % i == 0 || n % (i+2) == 0)
        {
            return false;
        }
        i += 6;
    }
    return true;
}

vector<int> rotations(vector<int> digits){
    vector<int> l {};

    for (int i = 0; i < digits.size() - 1; i++)
    {
        digits.insert(digits.begin(), digits.back());
        digits.pop_back();
        string tmp {};
        for (int i: digits)
            tmp += to_string(i);
        l.push_back(stoi(tmp));
    }

    return l;
}

bool is_circular_prime(int p){
    if (p == 2)
        return true;

    vector<int> digits {};
    int tmp {};

    for (char c: to_string(p)){
        tmp = c - '0';
        if (tmp % 2 == 0)
            return false;
        digits.push_back(tmp);
    }
    
    vector<int> combinations = rotations(digits);

    for (int x: combinations){
        if (! is_prime(x))
            return false;
    }

    return true;
}


int main() {  
    int s {};

    for (int i = 0; i < 1000000; i++)
    {
        if (is_prime(i) && is_circular_prime(i))
            s += 1;
    }

    cout << s << endl;

    return 0;
}